Objects/memoryobject.c

```diff
@@ -3160,6 +3162,133 @@ static PyMethodDef memory_methods[] = {
     {NULL,          NULL}
 };
 
+/**************************************************************************/
+/*                          Memoryview Iterator                           */
+/**************************************************************************/
+
+typedef struct {
+    PyObject_HEAD
+    Py_ssize_t it_index;
+    PyMemoryViewObject *it_seq; // Set to NULL when iterator is exhausted
+    Py_ssize_t length;
+    const char *fmt;
+} memoryiterobject;
+
+
+static void memoryiter_dealloc(memoryiterobject *);
+static int memoryiter_traverse(memoryiterobject * , visitproc, void *);
+static PyObject *memoryiter_next(memoryiterobject *);
+static PyObject *memory_iter(PyObject *);
+
+
+static PyTypeObject PyMemoryIter_Type = {
+    PyVarObject_HEAD_INIT(&PyType_Type, 0)
+    .tp_name = "memory_iterator",
+    .tp_basicsize = sizeof(memoryiterobject),
+    // methods
+    .tp_dealloc = (destructor)memoryiter_dealloc,
+    .tp_getattro = PyObject_GenericGetAttr,
+    .tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_GC,
+    .tp_traverse = (traverseproc)memoryiter_traverse,
+    .tp_iter = PyObject_SelfIter,
+    .tp_iternext = (iternextfunc)memoryiter_next,
+};
+
+static void
+memoryiter_dealloc(memoryiterobject *it)
+{
+    _PyObject_GC_UNTRACK(it);
+    Py_XDECREF(it->it_seq);
+    PyObject_GC_Del(it);
+}
+
+static int
+memoryiter_traverse(memoryiterobject *it, visitproc visit, void *arg)
+{
+    Py_VISIT(it->it_seq);
+    return 0;
+}
+
+static char *
+lookup_value(Py_buffer *view, char *ptr, Py_ssize_t index)
+{
+    Py_ssize_t nitems = view->shape[0];
+
+    // handle negative lookups (i.e [-2])
+    if (index < 0) {
+        index += nitems;
+    }
+    if (index < 0 || index >= nitems) {
```

Comment:
Since your only caller already checks that the index is within bounds, you could skip the `index >= nitems` check as well. And then you don't even need the `nitems` variable.

Reply:
Honest mistake - was reading lookup_dimension() and forgot that I didn't need those extra checks coming from memoryiter_next()